File: website/functions/src/models/card.ts
import {CardColor} from '../types/card-color';

export class Card {
  constructor(
    private color: CardColor,
    private value: number,
    private diff: number,
  ) {}

  static parse(data: any): Card {
    return new Card(
        data['color'],
        data['value'],
        data['diff'],
    );
  }

  static generate(amount: number, card: Card) {
    const result = [];

    for (let i = 0; i < amount; i++) {
      result.push(card);
    }

    return result;
  }

  static all() {
    const blue = [
      ...this.generate(3, new Card('blue', 1, 1)), // 1
      ...this.generate(3, new Card('blue', 2, 1)), // 2
      ...this.generate(3, new Card('blue', 3, 1)), // 3
      ...this.generate(3, new Card('blue', 4, 1)), // 4
      ...this.generate(2, new Card('blue', 5, 1)), // 5
      ...this.generate(2, new Card('blue', 6, 1)), // 6
      ...this.generate(2, new Card('blue', 7, 1)), // 7
      ...this.generate(2, new Card('blue', 8, 1)), // 8
      ...this.generate(2, new Card('blue', 9, 1)), // 9
      ...this.generate(2, new Card('blue', 10, 1)), // 10
    ];

    const yellow = [
      ...this.generate(2, new Card('yellow', 1, 2)), // 1
      ...this.generate(2, new Card('yellow', 2, 2)), // 2
      ...this.generate(2, new Card('yellow', 3, 2)), // 3
      ...this.generate(2, new Card('yellow', 4, 2)), // 4
      ...this.generate(3, new Card('yellow', 5, 2)), // 5
      ...this.generate(3, new Card('yellow', 6, 2)), // 6
      ...this.generate(3, new Card('yellow', 7, 2)), // 7
      ...this.generate(3, new Card('yellow', 8, 2)), // 8
      ...this.generate(2, new Card('yellow', 9, 2)), // 9
      ...this.generate(2, new Card('yellow', 10, 2)), // 10
    ];

    const red = [
      ...this.generate(3, new Card('red', 1, 3)), // 1
      ...this.generate(3, new Card('red', 2, 3)), // 2
      ...this.generate(3, new Card('red', 3, 3)), // 3
      ...this.generate(2, new Card('red', 4, 3)), // 4
      ...this.generate(2, new Card('red', 5, 3)), // 5
      ...this.generate(2, new Card('red', 6, 3)), // 6
      ...this.generate(2, new Card('red', 7, 3)), // 7
      ...this.generate(2, new Card('red', 8, 3)), // 8
      ...this.generate(3, new Card('red', 9, 3)), // 9
      ...this.generate(3, new Card('red', 10, 3)), // 10
    ];

    return [
      ...blue,
      ...yellow,
      ...red,
    ];
  }

  public json() {
    return {
      color: this.color,
      value: this.value,
      diff: this.diff,
    };
  }
}

File: server/lib/domain/models/card.dart
import 'package:undervoltage/domain/json/game/json_card.dart';
import 'package:undervoltage/domain/types/card_color.dart';

class Card {
  final String id;
  final CardColor color;
  final int diff;
  final int value;

  const Card({
    required this.id,
    required this.color,
    required this.diff,
    required this.value,
  });

  JsonCard get json => JsonCard(
        id: id,
        color: color,
        diff: diff,
        value: value,
      );

  bool canAccept(Card card) {
    final int value1 = _normalize(value + diff);
    final int value2 = _normalize(value - diff);

    return (card.value == value1) || (card.value == value2);
  }

  int _normalize(int value) {
    if (value > 10) {
      return value - 10;
    } else if (value < 1) {
      return value + 10;
    } else {
      return value;
    }
  }

  static List<Card> generate(
    int amount,
    CardId cardId,
    CardColor color,
    int value,
  ) {
    final List<Card> result = [];

    for (int i = 0; i < amount; i++) {
      result.add(Card(
        id: cardId.next.toString(),
        color: color,
        diff: color.diff,
        value: value,
      ));
    }

    return result;
  }

  @override
  String toString() => 'ID=$id,COLOR=${color.name},DIFF=$diff,VALUE=$value';

  static List<Card> all() {
    final CardId cardId = CardId(1);

    final List<Card> blue = [
      ...generate(3, cardId, CardColor.blue, 1),
      ...generate(3, cardId, CardColor.blue, 2),
      ...generate(3, cardId, CardColor.blue, 3),
      ...generate(3, cardId, CardColor.blue, 4),
      ...generate(2, cardId, CardColor.blue, 5),
      ...generate(2, cardId, CardColor.blue, 6),
      ...generate(2, cardId, CardColor.blue, 7),
      ...generate(2, cardId, CardColor.blue, 8),
      ...generate(2, cardId, CardColor.blue, 9),
      ...generate(2, cardId, CardColor.blue, 10),
    ];

    final List<Card> green = [
      ...generate(2, cardId, CardColor.green, 1),
      ...generate(2, cardId, CardColor.green, 2),
      ...generate(2, cardId, CardColor.green, 3),
      ...generate(2, cardId, CardColor.green, 4),
      ...generate(3, cardId, CardColor.green, 5),
      ...generate(3, cardId, CardColor.green, 6),
      ...generate(3, cardId, CardColor.green, 7),
      ...generate(3, cardId, CardColor.green, 8),
      ...generate(2, cardId, CardColor.green, 9),
      ...generate(2, cardId, CardColor.green, 10),
    ];

    final List<Card> red = [
      ...generate(3, cardId, CardColor.red, 1),
      ...generate(3, cardId, CardColor.red, 2),
      ...generate(3, cardId, CardColor.red, 3),
      ...generate(2, cardId, CardColor.red, 4),
      ...generate(2, cardId, CardColor.red, 5),
      ...generate(2, cardId, CardColor.red, 6),
      ...generate(2, cardId, CardColor.red, 7),
      ...generate(2, cardId, CardColor.red, 8),
      ...generate(3, cardId, CardColor.red, 9),
      ...generate(3, cardId, CardColor.red, 10),
    ];

    return [
      ...blue,
      ...green,
      ...red,
    ];
  }
}

class CardId {
  int _current;

  CardId(this._current);

  int get next => _current++;
}
